plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintCheck.kt

```diff
@@ -0,0 +1,106 @@
+package org.jlleitschuh.gradle.ktlint
+
+import net.swiftzer.semver.SemVer
+import org.gradle.api.DefaultTask
+import org.gradle.api.GradleException
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.FileTree
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Console
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.Nested
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.SkipWhenEmpty
+import org.gradle.api.tasks.TaskAction
+import org.jlleitschuh.gradle.ktlint.reporter.KtlintReport
+import org.jlleitschuh.gradle.ktlint.reporter.ReporterType
+import java.io.File
+
+open class KtlintCheck : DefaultTask() {
+
+    @get:Classpath
+    val classpath: ConfigurableFileCollection = project.files()
+
+    @get:Internal
+    val sourceDirectories: ConfigurableFileCollection = project.files()
+
+    @get:SkipWhenEmpty
+    @get:PathSensitive(PathSensitivity.RELATIVE)
+    @get:InputFiles
+    val sources: FileTree = sourceDirectories.asFileTree.matching { it.include("**/*.kt") }
+
+    @get:Input
+    val verbose: Property<Boolean> = booleanProperty()
+    @get:Input
+    val debug: Property<Boolean> = booleanProperty()
+    @get:Input
+    val android: Property<Boolean> = booleanProperty()
+    @get:Input
+    val ignoreFailures: Property<Boolean> = booleanProperty()
+    @get:Console
+    val outputToConsole: Property<Boolean> = booleanProperty()
+
+    @get:Internal
+    val reports: Map<ReporterType, KtlintReport> = ReporterType.values().map {
+        it to KtlintReport(project.objects.property(Boolean::class.javaObjectType).apply { set(false) }, it, newOutputFile())
+    }.toMap()
+
+    @get:Nested
+    val enabledReports
+        get() = reports.filterValues { it.enabled.get() }
+
+    @TaskAction
+    fun lint() {
+        val reportsToProcess = enabledReports.values
+        val ktlintVersion = determineKtlintVersion(classpath)
+        ktlintVersion?.let { version ->
+            if (version < SemVer(0, 20, 0)) {
+                reportsToProcess.forEach {
+                    val reportOutputPath = it.outputFile.get().asFile.absolutePath
+                    if (reportOutputPath.contains(" ")) {
+                        throw GradleException(
+                                "The output path passed `$reportOutputPath` contains spaces. Ktlint versions less than 0.20.0 do not support this.")
+                    }
+                }
+            }
+        }
+        project.javaexec {
+            it.classpath = classpath
+            it.main = "com.github.shyiko.ktlint.Main"
+            it.args(sourceDirectories.map { "${it.path}/**/*.kt" })
```

Comment:
Include `*.kts` incase someone is linting a `buildSrc` project with script based plugins.

Reply:
Do `*.kts` files lint well?

Reply:
@wolfs according to https://github.com/shyiko/ktlint/issues/19 since `0.3.0` release

Reply:
Making that change sounds orthogonal to this PR.
Even though it's a very interesting one :)